tests/test_code_references.py

```diff
@@ -200,3 +199,118 @@ def x(my_class):
             logging.debug("DEBUG log")
 
     assert get_references(x) == ["val", logging.info, logging.debug]
+
+
+def test_complex_function():
+    import inspect
+    import warnings
+    from bionic.code_hasher import CodeHasher, TypePrefix
+    from bionic.utils.reload import is_internal_file
+
+    hasher = CodeHasher(False)
+    # TODO: This makes me realize that maybe we should dedup the results.
+    assert get_references(hasher._ingest) == [
```

Comment:
If this test finds a bug in `CodeHasher`, then fixing the bug will also change the test, right?

Reply:
Changing `CodeHasher` code will likely change the returned references in this test too. Even if we don't add any new references, we will probably end up with a different order of the references. 

I tried a few methods from our test code as well (like the random tests), but it was referencing internal pytest methods and it was easier to use a complex method from Bionic code instead. I can always use a different method (replacing this or in addition to this) if you have something in mind.

Instead of just copying the returned references, I wrote all the references myself and then checked that the returned references were the same. This way I was able to make sure that we didn't miss a reference. I did add a TODO in this test which points that we might miss references from a conditional block statement. That's on my top of list of things to follow up on.

Reply:
Oh and if your concern is that the test is testing it's own source code, you shouldn't be worried about that. We are testing the logic for `get_code_references` by calling it on `CodeHasher`, which is separate from `get_code_references`.

Reply:
Oh right, I was wrong to think that changing CodeHasher would affect this test.

That said, this does seem to have the potential to introduce a lot of annoying busywork whenever we change the code of `CodeHasher`. I don't think this needs to block this PR though.